File: c/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length) {
  if (arr == NULL || length == 0)
    return NULL;

  int start = 0;
  int end = length;
  while (start < end) {
    int mid = (int)((end - start) / 2) + start;
    if (arr[mid] == value) {
      return arr + mid;
    } else if (arr[mid] > value) {
      end = mid;
    } else {
      start = mid + 1;
    }
  }
  return NULL;
}

File: dart/binary_search.dart
import 'package:binary_search/value_not_found_exception.dart';

class BinarySearch {
  final List<int> arr;

  BinarySearch(this.arr);

  int find(int target) {
    int lft = 0;
    int rht = arr.length - 1;
    while (lft <= rht) {
      final m = lft + (rht - lft) ~/ 2;
      if (arr[m] == target) {
        return m;
      }
      if (arr[m] < target) {
        lft = m + 1;
      } else {
        rht = m - 1;
      }
    }
    throw ValueNotFoundException('not found');
  }
}
